fs/socket/accept.c

```diff
@@ -38,10 +38,62 @@
 #include <nuttx/cancelpt.h>
 #include <nuttx/net/net.h>
 #include <nuttx/fs/fs.h>
+#include <nuttx/mm/kmap.h>
 #include <arch/irq.h>
 
+#include "sched/sched.h"
 #include "fs_heap.h"
 
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: map_user_mem
+ *
+ * Description:
+ *   Map userspace memory to continuous kernel virtual memory using
+ *   `kmm_map_user()`.
+ *
+ * Input Parameters:
+ *   uaddr    The user address to map. Can be NULL, in which case NULL will
+ *            be returned through `p_kaddr`.
+ *   size     The size of the mem region to map
+ *   p_kaddr  Pointer to the kernel address to return. Cannot be NULL.
+ *
+ * Returned Value:
+ *  Returns 0 (OK) on success, or:
+ *    -ENOMEM
+ *      If there is not enough free memory to create the mapping.
+ *    -EFAULT
+ *      If `uaddr` points to memory not belonging to the user address space.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_MM_KMAP
+static int map_user_mem(FAR void *uaddr, size_t size, void **p_kaddr)
```

Comment:
If the purpose of this function is to test if user has provided a valid user address (and not trying to give a kernel address for attacks), then it is unnecessary. The user given pointer checks should be more comprehensive i.e. on the system call layer for ALL SYSTEM CALLS (which we don't have yet). Regardless, this is for sure not the correct position to do such checks.

Reply:
The purpose of the function is to map user memory to kernel addressable memory.

In order to do that, `kmm_map_user()` also checks whether it is indeed the given task's memory it's trying to map, and if not, it returns non-NULL result. I reckon this is a desired error handling to do in this situation. 

AFAIK there is no legitimate use case for passing any other kind of memory (other task's, or kernel's address space) to this function - with `CONFIG_BUILD_KERNEL=y`, which `CONFIG_MM_KMAP` depends on.

Regarding the syscall layer, the function itself (along with others) might indeed be placed in a more accessible unit, and be exported to all syscall handlers. The actual checking using such functions though, not so easy; perhaps the first level pointers through some metadata (e.g. argN is a pointer to size `<sz>`), but nested types (structs with pointers etc) would be more tricky. Btw, Linux does manual checking in a very similar fashion to this PR (see `accept()`'s [invocation](https://github.com/torvalds/linux/blob/eabcdba3ad4098460a376538df2ae36500223c1e/net/socket.c#L1951) and [body](https://github.com/torvalds/linux/blob/eabcdba3ad4098460a376538df2ae36500223c1e/net/socket.c#L274-L299) of `move_addr_to_user()`). Not saying we can't do better, just that it is tricky.

Bottomline:
 - if you want to see the `EFAULT` checking removed please let me know what to do with this case (nothing else makes sense to me right now but I'll do anything to see it merged).
 - if you want to see `map_user_mem()` in some place else, please name it and I'll put it there along with any other changes you request.

Reply:
Thanks for the thorough explanation. I'm fine with it as long as there is a clear motivation for it.